osquery/core/system.cpp

```diff
@@ -193,11 +182,22 @@ Status getEphemeralUUID(std::string& ident) {
 }
 
 Status getHostUUID(std::string& ident) {
+  static bool checked_hardware_uuid = false;
+  if (!checked_hardware_uuid) {
+    // Attempt to get the hardware UUID from the system.
+    std::string hardware_uuid = osquery::generateHostUUID();
+    // If it's valid, store it in the database.
+    if (!hardware_uuid.empty() && !isPlaceholderHardwareUUID(hardware_uuid)) {
+      setDatabaseValue(kPersistentSettings, "host_uuid_v3", hardware_uuid);
+    }
+    checked_hardware_uuid = true;
+  }
```

Comment:
Does this need some locking logic for potential access from multiple threads?

Reply:
Good call.  This specific code is idempotent, but the method as a whole could have a race condition where two threads end up generating unique UUIDs.  I think this was probably an issue in the existing code too!  Added a lock here so only one thread will attempt retrieving and/or generating a UUID at a time.  This function should (race conditions aside) only be called once per process start, so the lock is inexpensive.